engine/runtime-compiler/src/main/java/org/enso/compiler/pass/lint/unusedimports/UnusedImportsRemover.java

```diff
@@ -0,0 +1,174 @@
+   *     would otherwise be lost. Note that if the statement was {@code from project.A.B import C,
+   *     D, E}, we want to replace it with something like {@code from project.A.B import C, D} and
+   *     not with {@code from Standard.Base.A.B import C, D}
+   * @return
+   */
+  private static LineReplacement replacementForUnusedSymbols(
+      UnusedSymbolsFromImport unusedSymsImp,
+      Import.Module impIr,
+      int lineIdx,
+      String importStatement) {
+    assert impIr.onlyNames().isDefined();
+    var onlyNames = impIr.onlyNames().get();
+    var unusedSyms = unqualified(asJava(unusedSymsImp.unusedSymbols()));
+    assert unusedSyms.size() <= onlyNames.size();
+    var usedSyms = onlyNames.filterNot(onlyName -> unusedSyms.contains(onlyName.name()));
+    if (usedSyms.isEmpty()) {
+      // Remove the line - no used symbols
+      return new LineReplacement(lineIdx, null);
+    } else {
+      var pat = Pattern.compile("from\\s+(.+)\\s+import(.+)$");
+      var matcher = pat.matcher(importStatement);
+      var found = matcher.find();
+      assert found;
+      var module = matcher.group(1);
+      var usedSymsStr = usedSyms.map(Name.Literal::name).mkString(", ");
+      var repl = "from " + module + " import " + usedSymsStr;
+      return new LineReplacement(lineIdx, repl);
+    }
+  }
+
+  /**
+   * @param replacements Locations sorted in reverse order, so they can be removed in one pass
+   */
+  private static void replaceLines(LazyFile lazyFile, List<LineReplacement> replacements) {
+    assert !replacements.isEmpty();
+    var oldLines = lazyFile.lines();
+    var newLines = new ArrayList<>(oldLines);
+    for (var replacement : replacements) {
+      var oldLinesIdx = replacement.lineIdx - 1;
+      var oldLine = oldLines.get(oldLinesIdx);
+      if (replacement.replacement != null) {
+        LOGGER.trace(
+            "Replacing line [{}] '{}' with '{}'", oldLinesIdx, oldLine, replacement.replacement);
+        newLines.set(oldLinesIdx, replacement.replacement);
+      } else {
+        LOGGER.trace("Removing line [{}] '{}'", oldLinesIdx, oldLine);
+        newLines.remove(oldLinesIdx);
+      }
+    }
+    try {
+      Files.write(lazyFile.path, newLines);
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  private static List<String> unqualified(List<String> list) {
+    return list.stream().map(s -> s.substring(s.lastIndexOf('.') + 1)).toList();
+  }
+
+  private static List<LineReplacement> reverseSort(List<LineReplacement> locations) {
+    return locations.stream()
+        .sorted((l1, l2) -> Integer.compare(l2.lineIdx(), l1.lineIdx()))
+        .toList();
+  }
+
+  /**
+   * @param lineIdx
+   * @param replacement If null, the line should be removed completely
+   */
+  private record LineReplacement(int lineIdx, String replacement) {}
+
+  private static final class LazyFile {
```

Comment:
It might be useful to use `LazyFile` via `for ( : )` loop:
```suggestion
  private static final class LazyFile implements Iterable<String> {
```

Reply:
`LazyFile.lines` are accessed randomly, for example from https://github.com/enso-org/enso/blob/ce066d1855d850c641d9327886e384e7025e0823/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/lint/unusedimports/UnusedImportsRemover.java#L64, so I would rather keep it this way.

Reply:
> randomly ... not sequentially

Then it'd be `class LazyFile extends AbstractList<String>`... The rest of the code could reference just `List<String>`.

But it is all private, so these changes wouldn't have a visible architectonic effect.... so we can let it be.